scripts/proteinInteractionEBI/parse_ebi_test.py

```diff
@@ -0,0 +1,123 @@
+ # Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''Test for parse_ebi.py.
+Run "python3 parse_ebi_test.py"
+'''
+import copy
+import unittest
+import parse_ebi
+
+
+
+TEST_TEXT = '''[Term]
+id: MI:0001
+name: interaction detection method
+def: "Method to determine the interaction." [PMID:14755292]
+
+[Term]
+id: MI:0045
+name: experimental interaction detection
+def: "Methods based" [PMID:14755292]
+is_a: MI:0001 ! interaction detection method
+
+[Term]
+id: MI:0401
+name: biochemical
+def: "The application" [PMID:14755292]
+is_a: MI:0045 ! experimental interaction detection'''
+
+CONST_ID_TO_CLASS_NAME = {'MI:0001': 'InteractionDetectionMethod',
+                          'MI:0045': 'ExperimentalInteractionDetection', 'MI:0401': 'Biochemical'}
+CONST_ID_TO_NODE = {}
+CONST_ID_TO_NODE_NO_RELATION = {}
+for key in ['MI:0001', 'MI:0045', 'MI:0401']:
+    CONST_ID_TO_NODE[key] = parse_ebi.Node(key)
+    CONST_ID_TO_NODE_NO_RELATION[key] = parse_ebi.Node(key)
+CONST_ID_TO_NODE['MI:0001'].child_list.append(CONST_ID_TO_NODE['MI:0045'])
+CONST_ID_TO_NODE['MI:0045'].parent_list.append(CONST_ID_TO_NODE['MI:0001'])
+CONST_ID_TO_NODE['MI:0045'].child_list.append(CONST_ID_TO_NODE['MI:0401'])
+CONST_ID_TO_NODE['MI:0401'].parent_list.append(CONST_ID_TO_NODE['MI:0045'])
+
+def get_file_terms(file):
+    "Ruturns a list of text blocks."
+    file_terms = file.split('\n\n')
+    file_terms = [term_text.split('\n') for term_text in file_terms
+                  if term_text.startswith('[Term]')]
+    return file_terms
+
+CONST_FILE_TERMS = get_file_terms(TEST_TEXT)
+
+CONST_INTERACTION_TYPE_ID_SET = set(['MI:0045', 'MI:0401'])
+
+class TestParseEbi(unittest.TestCase):
```

Comment:
These are tests are ok. But for future changes, you could simplify the amount of effort you put in testing and improve the end-to-end-ness of tests by structuring it as follows.

Make your core function, which you call from main(), like this:

   parse_ebi(input_file1, input_file2, ... , output_file1, output_file2)

Then you can prep sample input files based on a tiny subset of actual data, and prep expected output files.

Your test then will then do something like this:

def test_end_to_end(self):
  parse_ebi('testdata/ebi_input1.txt', 'testdata/ebi_input2.txt', ...., 'ebi_actual1.txt', 'ebi_actual2.txt')
  // assert that content of ebi_actual1.txt matches testdata/ebi_expected1.txt, and so on.

Reply:
I see! I was testing the important functions separately here. The end-to-end way is definitely easier. Will do it next time :)